### PR TITLE
bpo-45573: Introduce extension module flags in Makefile (GH-29594)

### PR DESCRIPTION
**Pull request by tiran**
``configure`` now uses a standardized format to forward state, compiler
flags, and linker flags to ``Makefile``, ``setup.py``, and
``Modules/Setup``. ``makesetup`` use the new variables by default if a
module line does not contain any compiler or linker flags. ``setup.py``
has a new function ``addext()``.

For a module ``egg``, configure adds:

* ``MODULE_EGG`` with value yes, missing, disabled, or n/a
* ``MODULE_EGG_CFLAGS``
* ``MODULE_EGG_LDFLAGS``

``Makefile.pre.in`` may also provide ``MODULE_EGG_DEPS`` that lists
dependencies such as header files and static libs.

Signed-off-by: Christian Heimes <christian@python.org>

https://bugs.python.org/issue45573

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -433,6 +382,74 @@ def __init__(self, dist):
...
+            self.extensions.append(ext)
+        elif state == "disabled":
+            self.disabled_configure.append(ext.name)
+        elif state == "missing":
+            self.missing.append(ext.name)
+        elif state == "n/a":
+            # not available on current platform
+            pass
+        else:
+            # not migrated to MODULE_{name} yet.
```
Comment (erlend-aasland) [inline setup.py:402]: Perhaps print an info message?

File: setup.py [hunk 2]
```diff
@@ -433,6 +382,74 @@ def __init__(self, dist):
...
+
+        Distutils appends extra args to the compiler arguments. Some flags like
+        -I must appear earlier, otherwise the pre-processor picks up files
+        from system inclue directories.
+        """
+        upper_name = ext.name.upper()
+        # Parse compiler flags (-I, -D, -U, extra args)
+        cflags = sysconfig.get_config_var(f"MODULE_{upper_name}_CFLAGS")
+        if cflags:
+            for token in shlex.split(cflags):
```
Comment (erlend-aasland) [inline setup.py:418]: Ah, it would have been convenient to be able write `for token in shlex.split(cflags) if cflags:` :)

**Review by erlend-aasland — APPROVED** [review]
Great work! LG

File: setup.py (hunk 2 above)
Comment (ambv) [reply setup.py:418]: You can `for token in shlex.split(cflags) if cflags else ():` if you really want to but I find the boring vanilla form less clever which is a Good Thing.

File: setup.py (hunk 1 above)
Comment (tiran) [reply setup.py:402]: Good thinking, I plan to add an error message after we have ported all stdlib modules to new system.

**Pull request merged**